src/__tests__/applicants.int.test.ts

```diff
@@ -66,6 +66,53 @@ describe('/applicants', () => {
         );
     });
 
+    it('returns the expected applicant by externalId', async () => {
+      await db.query(`
+        INSERT INTO applicants (
+          external_id,
+          opted_in,
+          created_at
+        )
+        VALUES
+          ( '54280485', 'false', '2023-04-25 17:39:00+0000' ),
+          ( '167336', 'true', '2023-04-05 17:40:00+0000' );
+      `);
+      await agent
+        .get('/applicants?externalId=167336')
+        .set(authHeader)
+        .expect(
+          200,
+          [
+            {
+              createdAt: '2023-04-05T17:40:00.000Z',
+              externalId: '167336',
+              id: 2,
+              optedIn: true,
+            },
+          ],
+        );
+    });
+
+    it('returns empty list when no applicants found by externalId', async () => {
+      await db.query(`
+        INSERT INTO applicants (
+          external_id,
+          opted_in,
+          created_at
+        )
+        VALUES
+          ( '3581', 'false', '2023-08-25 08:38:00+0000' ),
+          ( '3583', 'true', '2023-08-25 08:39:00+0000' );
+      `);
```

Comment:
Can we use the `applicants.insertOne` script here instead of directly executing SQL?

Reply:
Technically, yes. But I prefer the test setup code to be directly readable and minimize use of production code in tests other than the code under test.

Reply:
Brevity helps readability! And I would encourage you to re-evaluate that preference: here, we are tightly coupling this test about a search endpoint with knowledge of the underlying structure of the database table. That coupling is inappropriate, as it means that changing the database table will involve changing this unrelated test. Code should have only one reason to change; in this case, the test code should only change if the search endpoint changes.

Reply:
Adding required columns to the table will require this test to change regardless. Adding optional columns will not require a change to this test regardless. So I think that's unavoidable, right?

Reply:
I do think that it would make a lot of sense to leverage the power of `tinypg` for more concise tests (same way we don't have write SQL in our production code).

I personally would probably default to just having one set of (tested?) `.sql` templates rather than having two copies of `insertApplicant` where one is in `src/db` and the other is in like `src/test/db`, but even those two places would be more DRY than writing raw SQL inline every time.

Thoughts on that approach?

Reply:
While this review comment seemed like a small change to an existing pull request, it really is opening a new issue regarding our test code practices. I am open to changing our test code practices based on shared goals of test code. I do not think we need to hash it out here.

Reply:
Mixing SQL code as strings in TypeScript code is bad practice. It requires duplication of code, with the all attending issues of making changes to that duplicated code. It makes it harder to read, as it's just a string, rather than clearly marked as SQL, so editors/IDEs are less able to parse it as SQL. It makes it harder to lint, as it's just a string. We're using tinypg to be able to avoid all those problems!

> Adding required columns to the table will require this test to change regardless. Adding optional columns will not require a change to this test regardless. So I think that's unavoidable, right?

That specific example would require a matching change everywhere, yes. However, other changes might not; for example, renaming a table or a required column - but not the associated script name / parameters - would silently succeed when calling the script but require changes to SQL referring to those table/column names directly.

More importantly, making either such change - and reviewing it - will be easier when there's only one calling pattern to update, rather than two. It should be sufficient to search for the name of the SQL script and update each call site, rather than also need to replicate changes in multiple duplicate copies of in-line SQL code.

As we move forward on linting SQL (#321), linting will only ever work on SQL files - mixing SQL code as strings in TypeScript code makes that SQL opaque to our tools.

> While the original review comment seemed like a small change to an existing pull request, it really is opening a new issue regarding our test code practices. I am open to changing our test code practices based on shared goals of test code. I do not think we need to hash it out here.
> 
> It appears the issue might have already been created here: https://github.com/PhilanthropyDataCommons/service/issues/331

That issue is actually for improving our tests beyond the change I've requested here; the link in #331 goes to [a comment on #320](https://github.com/PhilanthropyDataCommons/service/pull/320#discussion_r1183117521) where we're using `applicants.insertOne`. And, in fact, we did hash this out previously: https://github.com/PhilanthropyDataCommons/service/pull/265#discussion_r1144197166 . It was not a long discussion, and we obviously didn't convert all existing SQL-strings-in-test-code as part of that, but I'm also not asking you to convert everything here - just stop introducing new SQL strings.

Reply:
I just wanna +1 that while it would be good to eventually make this fix everywhere, the first step towards improving coding practice / eliminating the tech debt is to make sure *new* contributions adhere to the best practice!

(We already do it both ways, and I think @jasonaowen's points explain why this would be an improvement to inline SQL.)

Reply:
[2fa610e](https://github.com/PhilanthropyDataCommons/service/pull/315/commits/2fa610e17fd8a1b6cace8885d115e31fcded7840) rebased on main.

[6d02917](https://github.com/PhilanthropyDataCommons/service/pull/315/commits/6d029176773404a8e92d062c3f27e2d3bc3a9f6e) has the requested change.

TLDR: do not misinterpret the code change here as forever binding assent to a new direction for the other existing tests or future tests. If you have time, read on.

The key readability of tests to me is at the moment of test failure. Put another way, the guiding question is "when a test fails, is it clear why that test failed?" This question is not exactly present for production code so there can be slight differences in the way we write tests versus production code. DRY is important in all code and is the strongest argument I have heard so far for changing the test to be more like the tests changed in https://github.com/PhilanthropyDataCommons/service/pull/265.

I want tests to fail for the right reasons and not the wrong reasons. To that end, a reason I avoid a call to production code (the insertOne code) in the test is to avoid the test failing for some bug in the code not under test here (the insertOne code). If the insert failed using a call to `db.sql` instead, that would be more obviously a bug in the test.

I want tests to have strong assertions. I want those assertions to be simple. I don't want tests to have what Meszaros calls a "Mystery Guest". To those ends, I include all the fields in the row and assert all fields in the row, literally and straightforwardly.

While it is OK to re-use the same data across tests, provided those data are carefully designed, I also find it helpful to have distinct data per test such that if I needed to search for a value in a given (i.e. failing) test's data or search the tests for a given data value (i.e. the db state after some test failures) I would be able to see exactly where that value came from.

So balancing DRY against these other objectives, I lean toward these other objectives more heavily than toward DRY. I acknowledge that DRY is better than not DRY in most circumstances. I am still open to revisiting the excess  test code duplication issue separately from this pull request.

<details><summary>As an example of what I'd like to avoid, when changing the code to be more closely aligned with the changes in #265 (for the sake of getting this code merged), I got this test failure:</summary>

```TypeScript
    // Here is the new code
    it('returns the expected applicant by externalId', async () => {
      await db.sql('applicants.insertOne', {
        externalId: '54280485',
        optedIn: false,
      });
      await db.sql('applicants.insertOne', {
        externalId: '167336',
        optedIn: true,
      });
      await agent
        .get('/applicants?externalId=167336')
        .set(authHeader)
        .expect(
          200,
          [
            {
              createdAt: expectTimestamp,
              externalId: '167336',
              id: 2,
              optedIn: true,
            },
          ],
        );
    });
```

</details>

``` TypeScript
// Here is the test failure
 FAIL  src/__tests__/applicants.int.test.ts (7.216 s)
  ● /applicants › GET / › returns the expected applicant by externalId

    expected [
      {
        createdAt: StringMatching {
          '$$typeof': Symbol(jest.asymmetricMatcher),
          sample: /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/,
          inverse: false
        },
        externalId: '167336',
        id: 2,
        optedIn: true
      }
    ] response body, got [
      {
        id: 2,
        externalId: '167336',
        optedIn: true,
        createdAt: '2023-09-12T16:17:42.084Z'
      }
    ]
```

<details><summary>Now I need to figure out why this test fails. It isn't clear from the above.</summary>Immediately I am given pause by the expected `createdAt` value. It is obscure and obviously different from `'2023-09-12T16:17:42.084Z'`. If I set that aside, I look at the remaining fields and find that they match. The `externalId` matches, the `optedIn` matches, and the `id` matches. This test passed before. Hmm. Perhaps the order of keys in a map is now significant. I go back to the original code and move the `createdAt`. What if I rule out the order by re-arranging it to `id, externalId, optedIn, createdAt`? Still fails. OK it's not order. No, order was not significant before. Is the import of the timestamp matcher the same? Yes. Perhaps the regex pattern does not exactly match. No, the regex looks good. Ah, the `expect` is now broken in two in the code pushed in #265, as [noted by @slifty](https://github.com/PhilanthropyDataCommons/service/pull/265#discussion_r1149553517). If I change the `expect` statements to this, it passes:

```TypeScript
        .expect(200)
        .expect((res) => expect(res.body).toEqual(
// ...
```
</details>

The reason I took the trouble to push back on these comments is I would like an acknowledgment that the change is not without its own cost. I might be misinterpreting, but the previous two comments come across as saying "the current code is bad practice and the suggested change is good practice" rather than recognizing that there are reasons for each approach and tradeoffs involved. One programmer might weigh the balance in one way and another might weigh it another way. We might all agree to goals of concise, readable, consistent, DRY, correctly passing, and correctly failing tests, but the priority order of the goals and our judgments about the long-term consequences of the exact implementation toward those goals might be different.

It was also brought up that "we did hash this out previously", but the "we" there was @slifty and @jasonaowen. My approval of a pull request is a judgment that, on balance, the code is an improvement or neutral and not moving backwards. If @slifty and @jasonaowen agree to something related to that code, it does not forever bind the remaining team members to that agreement. Nor is full agreement of everything in that pull request required for approval of a pull request. It is a different situation when I am asked to change something and after reflection still come to the conclusion that I prefer the original. It is also a different situation when the request is obviously not just about a few lines of code in a pull request but about a change in direction for code outside the request.

I think the benefit of review is improvements to the code. Where those suggested improvements are obvious to the requester, the person who requested review makes them with little trouble. Where those improvements are not obvious, the person doing the implementation should be given deference unless it is a big deal. I don't see this as a crucial, big deal to change these few lines to be on balance something like a neutral or slightly negative change. So I made the requested change here for the sake of getting this code merged.